### PR TITLE
Remove global state from RoutingDsl

### PR DESCRIPTION
**Pull request by marcospereira**
## Purpose

Remove global state from `RoutingDsl`. This deprecates both RoutingDls's default constructor and `build` methods.

### CODE REVIEW

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala [hunk 1]
```diff
@@ -8,14 +8,14 @@ import router.RoutingDslBuilder
...
+  def load(context: Context): Application = {
     new MyComponents(context).application
   }
 }
 
 class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) {
-  lazy val router = Router.from {
-       RoutingDslBuilder.getRouter.asScala.routes
+  lazy val router: Router = Router.from {
+    injector.instanceOf[RoutingDslBuilder].getRouter.asScala().routes
```
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: Agree. Will change it.

**Comment by marcospereira** [comment]
> I don't see the point of keeping RouterBuilderHelper at all. We can just move the state into RoutingDsl, and have it accept a BodyParser[AnyContent] and JavaContextComponents.

The point was to have a smaller change and also keep the old (and now deprecated) way working. So users would still be able to use `new RoutingDsl().build()` (which is deprecated), but could also use the new DI way. Also, `RouterBuilderHelper` is mostly doing translation between Scala and Java types, which is more clearer if written in Scala.

Agree with the other points of your review.

**Comment by gmethvin** [comment]
`RouterBuilderHelper` is also `private[routing]`, and I think the actual public API is `build()`. So I think it should be no problem to change it. I guess you can keep the class as a helper, but it would be much more useful if the `RoutingDsl` took the dependencies in the constructor directly so it can be easily used from `BuiltInComponents`.

File: framework/src/play-java/src/main/java/play/routing/RoutingDsl.java [hunk 2]
```diff
@@ -157,11 +195,36 @@ public PathPatternMatcher match(String method, String pathPattern) {
...
+        );
         return RouterBuilderHelper.build(this);
     }
 
+    /**
+     * Create the router.
+     *
+     * @return the created router.
+     */
+    public play.routing.Router create() {
```
Comment (gmethvin) [inline framework/src/play-java/src/main/java/play/routing/RoutingDsl.java]: I would just keep the `build()` method and only warn on the empty constructor.

File: framework/src/play-java/src/main/java/play/routing/RoutingDsl.java [hunk 3]
```diff
@@ -70,9 +84,33 @@
...
 
     /**
+     * Construct a new builder.
+     *
+     * @deprecated Deprecated as of 2.6.0. Use an injected version instead.
+     *
+     * @see #RoutingDsl(PlayBodyParsers, JavaContextComponents)
+     */
+    @Deprecated
+    public RoutingDsl() {
```
Comment (gmethvin) [inline framework/src/play-java/src/main/java/play/routing/RoutingDsl.java]: Would it be possible to just set the `bodyParser` and `contextComponents` from the running application here, to avoid the null checks and custom logic later?

File: documentation/manual/working/javaGuide/advanced/routing/JavaRoutingDsl.md [hunk 4]
```diff
@@ -9,6 +9,10 @@ The DSL is provided by [`RoutingDsl`](api/java/play/routing/RoutingDsl.html).  S
 
 @[imports](code/javaguide/advanced/routing/JavaRoutingDsl.java)
 
+And then you need to use [[Dependency Injection|JavaDependencyInjection]] to get a `RoutingDsl` instance:
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/JavaRoutingDsl.md]: You don't need to use dependency injection in the sense shown in the example. You just need to pass the body parser and JavaContextComponents. Since we explain below how to do it with different strategies, I'd just show manually passing to the constructor like:

```java
Router router = new RoutingDsl(defaultBodyParser, javaContextComponents)
    .GET("/hello/:to").routeTo(to ->
            ok("Hello " + to)
    )
    .build();
```

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala [hunk 5]
```diff
@@ -4,18 +4,20 @@
...
+  def load(context: Context): Application = {
     new MyComponents(context).application
   }
 }
 
 class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) {
-  lazy val router = Router.from {
-       RoutingDslBuilder.getRouter.asScala.routes
+  lazy val router: Router = Router.from {
+    val routingDsl = new RoutingDsl(playBodyParsers, javaContextComponents)
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: It would be nice if we could use the most specific dependency here and pass `defaultBodyParser` instead. Maybe we need two different constructors.

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala [hunk 6]
```diff
@@ -4,18 +4,20 @@
...
 
 //#load
 class AppLoader extends ApplicationLoader {
-  def load(context: Context) = {
+  def load(context: Context): Application = {
     new MyComponents(context).application
   }
 }
 
 class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) {
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: If we wanted to make it really easy, we could provide a trait to mix in that provides the `RoutingDsl`:
```scala
trait RoutingDslComponents {
  def defaultBodyParser: BodyParser[AnyContent]
  def javaContextComponents: JavaContextComponents
  lazy val routingDsl: RoutingDsl = new RoutingDsl(defaultBodyParser, javaContextComponents)
}
```

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala [hunk 7]
```diff
@@ -4,18 +4,20 @@
...
-  lazy val router = Router.from {
-       RoutingDslBuilder.getRouter.asScala.routes
+  lazy val router: Router = Router.from {
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: I can't remember if there is a good reason for `Router.from` here, if `new RoutingDslBuilder(routingDsl).getRouter.asScala` is already a router. Could we remove it?

**Comment by gmethvin** [comment]
@wsargent Can you also take a look?

File: framework/src/play-java/src/main/scala/play/routing/RouterBuilderHelper.scala [hunk 8]
```diff
@@ -19,7 +16,29 @@ import scala.collection.JavaConverters._
...
+  }
+
+  private def app = {
+    // If testing an embedded application we may not have a Guice injector, therefore we can't rely on
+    // it to instantiate the default body parser, we have to instantiate it ourselves.
+    Play.privateMaybeApplication.get // throw exception if no current app
+  }
+}
+
+private[routing] trait ScalaRouterBuilderHelper {
```
Comment (marcospereira) [inline framework/src/play-java/src/main/scala/play/routing/RouterBuilderHelper.scala]: Agree. Will change.

File: framework/src/play-java/src/main/java/play/routing/RoutingDsl.java (hunk 3 above)
Comment (marcospereira) [inline framework/src/play-java/src/main/java/play/routing/RoutingDsl.java]: Agree. Will change.

**Comment by wsargent** [comment]
Looking...

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala (hunk 6 above)
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: Problem here is that, to do that, we will need to add `Play-Java` as a dependency to the root project. Besides that, I'm not sure if it is a good idea to encourage people to reuse the same `RoutingDsl` all over the place (since it is mutable/non thread safe).

File: documentation/manual/working/javaGuide/advanced/routing/JavaRoutingDsl.md (hunk 4 above)
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/routing/JavaRoutingDsl.md]: Yep. This will need some rewrite after the reviews made.

**Comment by marcospereira** [comment]
I'm "tagging" this PR as WIP. See the most recent commit: https://github.com/playframework/playframework/pull/6887/commits/cf99b90db388470f156891edcc8d2b3f7ac7b8e0

There are possible additional methods (some of `Server.forRouter`) to be deprecated.

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala (hunk 6 above)
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: Can't we just put `RoutingDslComponents` in play-java? And we can make `routingDsl` a def so a new one will be used each time.

Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: Done.

**Comment by marcospereira** [comment]
@gmethvin and @wsargent, this is ready to be reviewed again.

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala [hunk 9]
```diff
@@ -4,18 +4,19 @@
...
-  def load(context: Context) = {
+  def load(context: Context): Application = {
     new MyComponents(context).application
   }
 }
 
-class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) {
-  lazy val router = Router.from {
-       RoutingDslBuilder.getRouter.asScala.routes
+class MyComponents(context: Context) extends BuiltInComponentsFromContext(context) with RoutingDslComponents {
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: So I think it would be technically possible to do this in Java instead if we created a `RoutingDslComponentsFromContext` that mixed in `RoutingDslComponents` for you. Then we could do this:

```java
class MyComponents extends RoutingDslComponentsFromContext {
  public MyComponents(Context context) {
    super(context);
  }

  public Router getRouter() {
    routingDsl()
      .GET("/hello/:to").routeTo(to -> ok("Hello " + to))
      .build();
  }
}
```

WDYT? Not totally necessary for this PR but it would be nice to eliminate the need for Scala code.

File: documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java [hunk 10]
```diff
@@ -29,12 +28,11 @@
     @Test
     public void simple() throws Exception {
         //#simple
-        Server server = Server.forRouter(new RoutingDsl()
-            .GET("/hello/:to").routeTo(to ->
-                ok("Hello " + to)
-            )
-            .build()
-        );
+        Server server = Server.forRouter((bodyParser, contextComponents) -> new RoutingDsl(bodyParser, contextComponents)
```
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java]: I like it.

Maybe we can completely remove these new `(bodyParser, contextComponents) -> Router` and add only `RoutingDsl -> Router` versions. This would be more straightforward and simple for users. Also looks like a more stable API.

File: documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala (hunk 9 above)
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/routing/code/AppLoader.scala]: Looks like a pretty small change. Will do it at this PR.

File: documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java (hunk 10 above)
Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java]: Just remember why I've did that way.

`play.server.Server` is part of `Play-Server` subproject and `RoutingDsl` is part of `Play-Java` and there is no dependency between these subprojects. User's code can add both, but to declare a `RoutingDsl -> Router` we will have to add a _weird_ dependency between the subprojects here.

WDYT?

Comment (gmethvin) [inline documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java]: I think we should use `Server.forRouter(BuiltInComponents -> Router)` and make the user construct the `RoutingDsl` manually from the `BuiltInComponents`.

We could also see if it makes sense to move `RoutingDsl` to the core `play` library. But that seems like something to do in a separate PR. Now that we've split Java forms out it might not be so bad to move all of play-java back to play.

Comment (marcospereira) [inline documentation/manual/working/javaGuide/advanced/embedding/code/javaguide/advanced/embedding/JavaEmbeddingPlay.java]: Done.

**Comment by marcospereira** [comment]
Rebased against master and ready and commits squashed to better describe the change.

This could be reviewed again. :-)

File: framework/src/play-integration-test/src/test/java/play/routing/AbstractRoutingDslTest.java [hunk 11]
```diff
@@ -19,187 +19,209 @@
...
+    abstract RoutingDsl routingDsl();
+
+    abstract Router router(RoutingDsl routing);
 
     @Test
     public void noParameters() {
-        Router router = new RoutingDsl()
-                .GET("/hello/world").routeTo(() -> Results.ok("Hello world"))
-                .build();
+        RoutingDsl routing = routingDsl()
```
Comment (gmethvin) [inline framework/src/play-integration-test/src/test/java/play/routing/AbstractRoutingDslTest.java]: It seems like this pattern could be replaced with something like
```java
Router router = router(routingDsl ->
  routingDsl.GET("/hello/world").routeTo(() -> Results.ok("Hello world"))
)
```

File: framework/src/play-integration-test/src/test/java/play/routing/NewRoutingDslTest.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <https://www.lightbend.com>
+ */
+package play.routing;
+
+// Should be removed together with deprecated methods and constructor at RoutingDsl
+public class NewRoutingDslTest extends AbstractRoutingDslTest {
```
Comment (gmethvin) [inline framework/src/play-integration-test/src/test/java/play/routing/NewRoutingDslTest.java]: Maybe call this `GlobalStateRoutingDslTest` to make it clear.

Comment (gmethvin) [inline framework/src/play-integration-test/src/test/java/play/routing/NewRoutingDslTest.java]: It might also be useful to have another test that creates the `RoutingDsl` from components instead of using runtime dependency injection.

File: framework/src/play/src/main/scala/play/api/mvc/BodyParsers.scala [hunk 13]
```diff
@@ -698,6 +698,9 @@ trait PlayBodyParsers extends BodyParserUtils {
    */
   def default: BodyParser[AnyContent] = default(None)
```
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/api/mvc/BodyParsers.scala]: Maybe we should just rename this but it could be done in another PR

**Comment by marcospereira** [comment]
@gmethvin all done. :-)
